docs/provisioning.rst

```diff
@@ -0,0 +1,355 @@
+.. _provisioning
+
+Customizing the kernel's runtime environment
+============================================
+
+Kernel Provisioning
+~~~~~~~~~~~~~~~~~~~
+
+Introduced in the 7.0 release, Kernel Provisioning enables the ability
+for third parties to manage the lifecycle of a kernel's runtime
+environment. By implementing and configuring a *kernel provisioner*,
+third parties now have the ability to provision kernels for different
+environments, typically managed by resource managers like Kubernetes,
+Hadoop YARN, Slurm, etc. For example, a *Kubernetes Provisioner* would
+be responsible for launching a kernel within its own Kubernetes pod,
+communicating the kernel's connection information back to the
+application (residing in a separate pod), and terminating the pod upon
+the kernel's termination. In essence, a kernel provisioner is an
+*abstraction layer* between the ``KernelManager`` and today's kernel
+*process* (i.e., ``Popen``).
+
+The kernel manager and kernel provisioner relationship
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Prior to this enhancement, the only extension point for customizing a
+kernel's behavior could occur by subclassing ``KernelManager``. This
+proved to be a limitation because the Jupyter framework allows for a
+single ``KernelManager`` class at any time. While applications could
+introduce a ``KernelManager`` subclass of their own, that
+``KernelManager`` was then tied directly to *that* application and
+thereby not usable as a ``KernelManager`` in another application. As a
+result, we consider the ``KernelManager`` class to be an
+*application-owned entity* upon which application-specific behaviors can
+be implemented.
+
+Kernel provisioners, on the other hand, are contained within the
+``KernelManager`` (i.e., a *has-a* relationship) and applications are
+agnostic as to what *kind* of provisioner is in use other than what is
+conveyed via the kernel's specification (kernelspec). All kernel
+interactions still occur via the ``KernelManager`` and ``KernelClient``
+classes within ``jupyter_client`` and potentially subclassed by the
+application.
+
```

Comment:
Does the KernelManager or KernelClient API changes (additional (optional) methods, method signature change...) ? Is there any impact for current consumers of those APIs?

Maybe worth stating this in the documentation?

Reply:
Nothing changed on `KernelClient`.  As for `KernelManager` there were some changes to "internal" methods and I agree those kinds of changes should be documented (primarily calls to the Popen instance are replaced with calls to the provisioner instance).  As a result, only `KernelManager` subclass authors should be affected (and hopefully minimally).  However, I would suggest those be part of the 7.0 release docs (or whatever major release this falls in) rather than embedded in functional/dev docs pertaining to Kernel Provisioning itself.